man/man1/mkspiral.1
.TH MKSPIRAL 1NEMO "6 June 2024"

.SH "NAME"
mkspiral \- uniform-density test disk plus spiral in spherical potential

.SH "SYNOPSIS"
\fBmkspiral out=\fPsnapshot [parameters=values ...]

.SH "DESCRIPTION"
\fImkspiral\fP sets up a cold disk of test particles orbiting clockwise
in the gravitational potential of a user-supplied potential (in
\fIpotential(5NEMO)\fP format). In addition a spiral is added with
constant width in degrees along a galactic circle. For given ratio
of spiral/disk perturbation \fBa\fP and width \fBw\fP (now measured in
radians) the fraction of particles in the disk will be about
\fB pi / (pi + a.w.sqrt(2.pi))\fP for not too large \fBw\fP.
The inverse wavelenght of the spiral (related to the pitch angle)
will be positive for a \fItrailing spiral pattern\fP.
.PP
\fIsnapmass(1NEMO)\fP and \fIsnapcenter(1NEMO)\fP have to be used if
for some reason the user want finite mass particles and a perfectly
centered set of particles.
.PP
Note this particle distribution is only useful for display, e.g. spiral
structure in a Longitude-Velocity diagram, and should not be integrated
with programs like \fIhackcode3(1NEMO)\fP.
.PP
\fBmkspiral\fP is also used in the standard benchmark as to measure
disk I/O speed on a file just over 2GB.

.SH "PARAMETERS"
.so man1/parameters
.TP 25
\fBout=\fIout_file\fP
The data are written to this file in \fISnapShot(5NEMO)\fP 
format
.br
[no default].
.TP
\fBnbody=\fInum_bodies\fP
Number of disk particles
.br
[default: \fB2048\fP]
.TP
\fBpotname=\fIpotential_name\fP
name of the potential, see $NEMO/obj/potential for the current
repository. The user can supply his own, see potential(5NEMO).
.br
[default: \fBplummer\fP].
.TP
\fBpotpars=\fIpot_pars\fP
Paramaters to the user supplied potential. The number of parameters
depends on the potential supplied, the first parameter is reserved
for a pattern speed in future implementation.
.br
[default: not supplied, parameters as defined by potential(5)].
.TP
\fBpotfile=\fIpot_file\fP
Data_file(s) to the user supplied potential. The number of files depends
on the potential supplied.
.br
[default: not supplied, datafile(s) defined by potential(5)].
.TP
\fBrmin=\fImin_disk_radius\fP
Inner cutoff radius of test-particle disk.
.br
[Default: \fB0.0\fP].
.TP
\fBrmax=\fImax_disk_radius\fP
Outer cutoff radius of test-particle disk.
.br
[Default: \fB1.0\fP].
.TP
\fBmass=\fP
Total mass of the disk.
.br
[Default: 0]
.TP
\fBa=\fIratio\fP
Ratio of maximum spiral perturbation w.r.t. base disk.
.br
[default: \fB1\fP].
.TP
\fBk=\fIinv_rad\fP
Wavenumber of the spiral pattern, such that \fB2.pi.k.r\fP is the phase of
the spiral density maximum.  Positive \fB\fP means trailing spiral pattern.
(note that this is defined in term of \fBsign=\fP, see below)
.br
[default: \fB1\fP]. 
.TP
\fBw=\fIwidth\fP
Width of the spiral arm along a galactic circle. Depending on the next keyword
this width is either is degrees along a galactic circle if that is to be 
constant, or in length units if the width of the spiral is to be constant.
.br
[default: \fB15\fP].
.TP
\fBangular=t|f\fP
Is the width of the spiral defined in angular terms, in which case the
physical real width linearly increases with distance from the galactic
center, or is is constant as a function of radius.
.br
[default: \fBt\fP, i.e. defined in angular terms].
.TP
\fBseed=\fIrandom_seed\fP
Use random number seed. A zero seed will generate a system-time
dependant seed.
.br
[default: \fB0\fP].
.TP
\fBsign=-1|+1\fP
Sign of the angular moment vector. Positive means counter-clock wise 
rotation.
.br
[Default: 1]
.TP
\fBsigma=\fP\fIs1,s2,s3\fP
Velocity dispersion (isotropic) in the X-Y plane, plus optionally
an offset and exponential dropoff: s = s1 + s2 * exp(-r/s3).
.br
[Default: 0]
.TP
\fBsigmator=\fP
The ratio of the tangential to radial velocity dispersion. By default 1,
but if you need realistic values, our local solar neighbor measures about
0.5 but the best self-consistent (despite these models are NOT)
estimate can be set by using the value -1, in which case 0.5*kappa/omega
is used.
.br
[Default: 1]
.TP
\fBnmodel=\fP\fIn_models\fP
Number of output models.
.br
[Default: \fB1\fP].
.TP
\fBheadline=\fImessage\fP
Text headline for output file
.br
[default: not used].
.TP
\fBlinear=t|f\fP
Should the spiral be linear or logarithmic. Although spiral galaxies are
mostly logarithmic, for historic reasons the default here is linear.
.br
[Default: \fBt\fP]
.TP
\fBtest=\fItest-mode\fP
The mode for testing, an integer.
.br
0: gives the normal test disk;
.br
1: the 2D coordinates are uniform random (-1,1);
.br
2: the 2D coordinates are all set to 1.0;
.br
3: no initialization (everything 0)
.br
[Default: 0]

.SH "BENCHMARK"
This program has been used as benchmark to test writing >2GB files
(40 models of 1,000,000 particles creates a file just over this
magic limit) at a time when this could be critical:
.EX

    % time mkspiral s000 1000000 nmodel=40
    
.EE
Here is a sample on some 2002-class machines:
.nf
  U-sparc IIi/440 MHz:  161.30u 42.59s 5:04.85 66.8%  
  PentiumIV 2.2 GHz:    29.54u 10.68s 2:40.14 25.1% 
  PentiumIV 2.4 GHz:    26.46u 10.14s 1:37.71 37.4%  
  G5 2.0GHz:            28.400u 24.660s 1:05.41 81.1% (at no load of 0.0)
                        33.460u 44.950s 13:01.31 10.0%  (at full load of 2.0)
  P4/1.6 GHz:           69.940u 18.470s 4:50.64 30.4% (laptop)
  P4/2.6 GHz + SATA     27.400u 16.320s 0:52.13 83.8%   0+0k 0+0io 300pf+0w (at no load of 0.0)
                        27.790u 20.020s 1:52.22 42.6%   0+0k 0+0io 300pf+0w
  Ultra 155H            2.56user 1.26system 0:04.15elapsed 92%CPU

sum: 79.35u  3.72s 2:17.64 60.3%   P2400
tsf:  0.00u  7.59s 1:59.46  6.3%   P2400

sum: 13.610u 4.130s 1:10.12 25.2%  P2600/SATA
tsf:  0.010u 7.800s 1:09.74 11.1%  P2600/SATA

sum: 13.334u 2.915s 0:57.67 28.1%  P2800/SATA (Peter's asus )
tsf:  0.003u 4.696s 0:44.36 10.5%  P2800/SATA

sum: 68.63u 22.34s 2:31.54 60.0%  U440
tsf: 21.85u 30.73s 2:32.06 34.5%  U440

.fi
And here a more modern benchmark on a i7-8550U CPU processor with different modes:
.nf
   % time mkspiral $out 1000000 nmodel=40 seed=123 test=$test

   $test  $out  cpu:
   0      .     2.45u 0.47s 0:02.91  ultra 7 155H
   0      .     3.05u 1.09s          i5-1135G7
   0      .     4.15u 1.56s          i5-10xxx
   0      s0    6.21u 2.16s 0:08.38  full computation on a  i7-8550U CPU
   0      .     6.40u 0.69s 0:07.10   
   1      .     3.49u 0.68s 0:04.17  only using 4 xrandom's calls
   2      .     2.32u 0.71s 0:03.03  constant initialization
   2      .     1.43u 0.68s 0:02.12  - same , but with better caching barebody.h
   3      .     0.65u 0.56s 0:01.22  no initialization (using calloc)
.fi

.SH "SEE ALSO"

mkdisk(1NEMO), snapmass(1NEMO), snapcenter(1NEMO), hackcode3(1NEM0)

.SH "HISTORY"
.nf
.ta +1.25i +4.25i
13-mar-89	V1.0  created                   	PJT
6-apr-89	V1.1 seed=0 implemented          	PJT
7-apr-89	V1.2 angular keyword introduced  	PJT
26-feb-93	V1.4 new potential keywrds	PJT
16-sep-95	V1.5 added nmodel=, sign= ; radii random	PJT
26-may-02	added benchmark data for >2GB files	PJT
3-aug-09	V1.8 and documented forgotten stuff		PJT
11-aug-09	V1.9 added sigmator=	PJT
24-oct-2020	V1.10 added test=	PJT
.fi


src/nbody/init/mkspiral.c
/*
 * MKSPIRAL.C: set up a uniform-density test disk  with spiral perturbation
 *	in a spherical potential(5) - based on MKDISK
 *	
 *	original version: 13-Mar-89	Peter Teuben - Maryland
 *	 7-apr-90 V1.2 ???					PJT
 *      25-nov-90 V1.3 helpvec etc. 				PJT
 *	20-feb-92 V1.3a usage, nemo_main			PJT
 *	23-feb-93 V1.3b fixed potential() arglist bug		PJT
 *       6            c now writing history			PJT
 *      26-feb-93 V1.4  potname,potpars,potfile official potential keywords
 *	15-sep-95 V1.5  allow many models nmodel=, add sign=	PJT
 *                      radii now fully random fromrmin to rmax
 *	23-mar-97 V1.5b fixed for SINGLEPREC			pjt
 *      27-mar-97 V1.6  add mass= for optional total mass       pjt
 *                      but actually was still broken for SINGLEPREC
 *      10-mar-01 V1.7  allow sigma= to have 3 parameters
 *       8-sep-01       a   init_xrandom
 *       8-apr-03       b   timebit
 *      22-apr-09 V1.8  linear/logarithm spiral option
 *      11-aug-09 V1.9  add sigmator= to control anisotropy     PJT
 */

#include <stdinc.h>
#include <getparam.h>
#include <vectmath.h>
#include <filestruct.h>
#include <history.h>
#include <potential.h>
#include <spline.h>

#include <snapshot/snapshot.h>
#include <snapshot/body.h>
//#include <snapshot/barebody.h>
#include <snapshot/put_snap.c>

string defv[] = {
    "out=???\n		  output file name",
    "nbody=2048\n	  number of particles",
    "potname=plummer\n    name of background potential",
    "potpars=\n           optional parameters to potential",
    "potfile=\n           optional data file with potential",
    "rmin=0.0\n	          inner disk radius",
    "rmax=1.0\n		  outer cutoff radius",
    "mass=0\n             total mass of disk",
    "a=1\n                relative spiral perturbation w.r.t. disk",
    "k=1\n                spiral wavenumber  (> 0 trailing spirals)",
    "w=15\n               gaussian width of spiral - in degrees",
    "angular=t\n	  constant angular width",
    "seed=0\n		  random number seed",
    "sign=1\n             sign of angular momentum of disk",
    "sigma=0\n            velocity dispersion, plus optional central offset and exp scalelength",
    "sigmator=1\n         Ratio sigma_t/sigma_r:  use < 0 to compute it from kappa/2.omega",
    "nmodel=1\n           number of models",
    "headline=\n	  text headline for output ",
    "linear=t\n           Linear or Logarithmic spiral?",
    "test=0\n             0=real disk  1=low cpu garbage disk",
    "VERSION=1.10\n	  24-oct-2020 PJT",
    NULL,
};

string usage = "toy spiral density perturbation in a uniform disk";

string cvsid="$Id$";


local real rmin, rmax, sigma[3], sigmator;
local int  ndisk, nmodel, nsigma, sign;
local real SPa, SPk, SPw;     /* spiral parameters */
local real width;
local real totmass;
local bool Qlinear;
local int  testmode;

local Body *disk = NULL;

local proc potential;

extern double xrandom(double, double);
extern double grandom(double, double);

local void inittables(void);
local void writegalaxy(stream outstr);
local void testdisk0(int n);
local void testdisk1(int n);
local void testdisk2(int n);
local void testdisk3(int n);

void nemo_main(void)
{
    stream outstr;
    
    potential = get_potential(getparam("potname"),
                    getparam("potpars"), getparam("potfile"));
    rmin = getdparam("rmin");
    rmax = getdparam("rmax");
    ndisk = getiparam("nbody");
    nmodel = getiparam("nmodel");
    sign = getiparam("sign");
    nsigma = nemoinpr(getparam("sigma"),sigma,3);
    sigmator = getrparam("sigmator");
    totmass = getdparam("mass");
    Qlinear = getbparam("linear");
    testmode = getiparam("test");
    
    SPa = getdparam("a");
    SPk = - getdparam("k");	/* corrected for rot counter clock wise */
    if (sign < 0) SPk *= -1.0;  /* flip K if spinning the other way around */
    SPw = getdparam("w");
    if (getbparam("angular")) {
	width = -1.0;
	SPw /= 180 * PI;		/* convert degrees to radians */
    } else
        width = SPw;
    init_xrandom(getparam("seed"));
    if (sigmator<1) inittables();
    if (testmode > 0) warning("garbage test disk mode %d",testmode);

    outstr = stropen(getparam("out"), "w");
    put_history(outstr);
    if (hasvalue("headline"))
	set_headline(getparam("headline"));
    
    while (nmodel--) {
      switch (testmode) {
      case 0: testdisk0(nmodel); break; 
      case 1: testdisk1(nmodel); break; 
      case 2: testdisk2(nmodel); break;
      case 3: testdisk3(nmodel); break;
	
      default: break;
      }
      writegalaxy(outstr);
    }
    strclose(outstr);
    free(disk);
}

#define NTAB 256

local real rcir[NTAB];
local real vcir[4*NTAB];


local void inittables()
{

  int i, ndim=3;
  double pos_d[3], acc_d[3], pot_d, time_d = 0.0;

  pos_d[0] = pos_d[1] = pos_d[2] = 0.0;

  rcir[0] = vcir[0] = 0.0;
  for (i = 1; i < NTAB; i++) {
    pos_d[0] = rcir[i] = rmax * ((real) i) / (NTAB - 1);
    (*potential)(&ndim,pos_d,acc_d,&pot_d,&time_d);      /* get forces    */	
    vcir[i] = -acc_d[0] * rcir[i];
    if (vcir[i] < 0) error(" (inittables) force not inward bound at r=%g",rcir[i]);
    vcir[i] = sqrt(vcir[i]);
  }
  spline(&vcir[NTAB], &rcir[0], &vcir[0], NTAB);
}


/*
 * WRITEGALAXY: write galaxy model to output.
 */

void writegalaxy(stream outstr)
{
    real tsnap = 0.0;
    int bits = MassBit | PhaseSpaceBit | TimeBit;

    put_snap(outstr, &disk, &ndisk, &tsnap, &bits);
}

/*
 * TESTDISK: use forces due to a potential to make a uniform
 * density test disk.  
 */

void testdisk0(int n)
{
    Body *dp;
    real rmin2, rmax2, r_i, theta_i,vcir_i, mass_i, sigma_i;
    real uni, gau, unifrac, f;
    real cost, sint, Aoort, omega, kappa, sigma_t;
    vector acc_i;
    bool Qsigma = (sigma[0]>0.0 || sigma[1]>0.0);
    int i, ndim=NDIM;
    double pos_d[NDIM], acc_d[NDIM], pot_d, time_d = 0.0;
    

    if (disk == NULL) disk = (Body *) allocate(ndisk * sizeof(Body));
    rmin2 = rmin * rmin;
    rmax2 = rmax * rmax;
    mass_i = 1.0 / (real)ndisk;
    uni = PI;                            /* area of uniform */
    if (width<0)                    /* constant angle spiral */
        gau = SPa * SPw * sqrt(TWO_PI);    /* area of gauss - erf??? */
    else                            /* fxied width spiral */
        gau = SPa * atan(2*SPw/(rmin+rmax)) * sqrt(TWO_PI);
    unifrac = uni / (uni+gau);
    dprintf(1,"fraction of particles in disk = %f (%d)\n",unifrac,n);
    if (Qsigma > 0)  
      warning("hot disk, sigma=%g,%g,%g",sigma[0],sigma[1],sigma[2]);

    for (dp=disk, i = 0; i < ndisk; dp++, i++) {
	Mass(dp) = mass_i;
        r_i = sqrt(rmin2 + xrandom(0.0,1.0) * (rmax2 - rmin2));
        f = xrandom(0.0,1.0);			   /* pick random disk/spiral */
	if (f<unifrac)                                     /* disk particle   */
	    theta_i = xrandom(0.0, TWO_PI);	/* pick random angle          */
	else {                                           /* spiral particle   */
            if (width>0)                         /* variable angular width    */
                SPw = atan(width/r_i);              
            theta_i = grandom(0.0,SPw);		/* get gaussian angle         */
            f = xrandom(0.0,1.0);		/* pick which spiral arm      */
            if (f<0.5)                     
                theta_i += PI;
	    if (Qlinear)
	      theta_i -= SPk * r_i * TWO_PI;    /* positive SPk is trailing SP  */
	    else
	      theta_i -= SPk * log(r_i) * TWO_PI;    /* positive SPk is trailing SP  */
        }
        cost = cos(theta_i);
        sint = sin(theta_i);
	Phase(dp)[0][0] = pos_d[0] = r_i * cost;        /* set positions      */
	Phase(dp)[0][1] = pos_d[1] = r_i * sint;
	Phase(dp)[0][2] = pos_d[2] = 0.0;
        (*potential)(&ndim,pos_d,acc_d,&pot_d,&time_d);      /* get forces    */
        SETV(acc_i,acc_d);
	vcir_i = sqrt(r_i * absv(acc_i));
        dprintf(2,"r=%g vcir=%g acc=%g %g\n",r_i,vcir_i,acc_i[0],acc_i[1]);
        if (Qsigma) {
	    sigma_i = sigma[0] + sigma[1]*exp(-r_i/sigma[2]);
	    if (sigmator < 0) {
	      omega = vcir_i / r_i;
	      Aoort = -0.5 * (spldif(r_i, &rcir[0], &vcir[0], &vcir[NTAB], NTAB) - omega);
	      if (omega - Aoort < 0.0)
		dprintf(0,"R, omega, Aoort = %f %f %f\n", r_i, omega, Aoort);
	      kappa = 2 * sqrt(omega*omega - Aoort * omega);
	      sigma_t = 0.5 * sigma_i * kappa / omega;
	    } else
	      sigma_t = sigmator * sigma_i;
	    sigma_i = grandom(0.0,sigma_i);
	    sigma_t = grandom(0.0,sigma_t);
        } else
	  sigma_i = sigma_t = 0.0;
	/* sign > 0 makes for counter clock wise */
	Phase(dp)[1][0] = - sign * (vcir_i+sigma_t) * sint + sigma_i * cost;
	Phase(dp)[1][1] =   sign * (vcir_i+sigma_t) * cost + sigma_i * sint;
	Phase(dp)[1][2] = 0.0;
    }
}

/*
 * a garbage disk with as little CPU used as possible
 * meant to calibrate disk I/O factor
 */

void testdisk1(int n)
{
    Body *dp;
    int i;

    if (disk == NULL) disk = (Body *) allocate(ndisk * sizeof(Body));

    for (dp=disk, i = 0; i < ndisk; dp++, i++) {
        Mass(dp) = 1.0;
 	Phase(dp)[0][0] = xrandom(-1.0,1.0);
	Phase(dp)[0][1] = xrandom(-1.0,1.0);
	Phase(dp)[0][2] = 0.0;
	Phase(dp)[1][0] = xrandom(-1.0,1.0);
	Phase(dp)[1][1] = xrandom(-1.0,1.0);
	Phase(dp)[1][2] = 0.0;
    }
}


void testdisk2(int n)
{
    Body *dp;
    int i;

    if (disk == NULL) disk = (Body *) allocate(ndisk * sizeof(Body));

    for (dp=disk, i = 0; i < ndisk; dp++, i++) {
        Mass(dp) = 1.0;
 	Phase(dp)[0][0] = 1.0;
	Phase(dp)[0][1] = 1.0;
	Phase(dp)[0][2] = 0.0;
	Phase(dp)[1][0] = 1.0;
	Phase(dp)[1][1] = 1.0;
	Phase(dp)[1][2] = 0.0;
    }
}

void testdisk3(int n)
{
    if (disk == NULL) disk = (Body *) allocate(ndisk * sizeof(Body));
}
